File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryBasicNameMapper.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Binary
{
    using System.Diagnostics.CodeAnalysis;
    using System.Globalization;
    using System.Text;
    using Apache.Ignite.Core.Impl.Binary;
    using Apache.Ignite.Core.Impl.Common;

    /// <summary>
    /// Base binary name mapper implementation.
    /// </summary>
    public class BinaryBasicNameMapper : IBinaryNameMapper
    {
        /// <summary>
        /// The simple name instance.
        /// </summary>
        internal static readonly BinaryBasicNameMapper SimpleNameInstance = new BinaryBasicNameMapper
        {
            IsSimpleName = true
        };

        /// <summary>
        /// The full name instance.
        /// </summary>
        internal static readonly BinaryBasicNameMapper FullNameInstance = new BinaryBasicNameMapper();

        /// <summary>
        /// Gets or sets a value indicating whether this instance maps to simple type names.
        /// </summary>
        public bool IsSimpleName { get; set; }

        /// <summary>
        /// Gets or sets the prefix to be added to the full type name.
        /// For example, Java package names usually begin with <c>org.</c> or <c>com.</c>.
        /// <para />
        /// In combination with <see cref="NamespaceToLower"/>, we can map .NET type name <c>Apache.Ignite.Foo</c>
        /// to a corresponding Java type name <c>org.apache.ignite.Foo</c>, conforming to the naming conventions
        /// for both languages.
        /// </summary>
        public string NamespacePrefix { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this instance converts the namespace part to the lower case.
        /// For example, Java package names are usually lowercase.
        /// <para />
        /// In combination with <see cref="NamespacePrefix"/>, we can map .NET type name <c>Apache.Ignite.Foo</c>
        /// to a corresponding Java type name <c>org.apache.ignite.Foo</c>, conforming to the naming conventions
        /// for both languages.
        /// </summary>
        public bool NamespaceToLower { get; set; }

        /// <summary>
        /// Gets the type name.
        /// </summary>
        public string GetTypeName(string name)
        {
            IgniteArgumentCheck.NotNullOrEmpty(name, "typeName");

            var parsedName = TypeNameParser.Parse(name);

            if (parsedName.Generics == null)
            {
                // Generics are rare, use simpler logic for the common case.
                return GetTypeName(parsedName) + parsedName.GetArray();
            }

            return BuildTypeName(parsedName, new StringBuilder()).ToString();
        }

        /// <summary>
        /// Gets the field name.
        /// </summary>
        public string GetFieldName(string name)
        {
            return name;
        }

        /// <summary>
        /// Builds the type name.
        /// </summary>
        private StringBuilder BuildTypeName(TypeNameParser typeName, StringBuilder sb)
        {
            sb.Append(GetTypeName(typeName));

            var generics = typeName.Generics;

            if (generics != null && generics.Count > 0)  // Generics are non-null but empty when unbound.
            {
                sb.Append('[');

                var first = true;

                foreach (var genArg in generics)
                {
                    if (first)
                    {
                        first = false;
                    }
                    else
                    {
                        sb.Append(',');
                    }

                    sb.Append('[');

                    BuildTypeName(genArg, sb);

                    sb.Append(']');
                }

                sb.Append(']');
            }

            sb.Append(typeName.GetArray());

            return sb;
        }

        /// <summary>
        /// Gets the type name from the parser.
        /// </summary>
        [SuppressMessage("Microsoft.Globalization", "CA1308:NormalizeStringsToUppercase",
            Justification = "Not applicable, lower case is required.")]
        private string GetTypeName(TypeNameParser name)
        {
            if (IsSimpleName)
                return name.GetName();

            var fullName = NamespaceToLower && name.HasNamespace()
                ? name.GetNamespace().ToLower(CultureInfo.InvariantCulture) + name.GetName()
                : name.GetNameWithNamespace();

            return NamespacePrefix + fullName;
        }
    }
}

File: modules/core/src/main/java/org/apache/ignite/binary/BinaryBasicNameMapper.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.binary;

import org.apache.ignite.internal.util.typedef.internal.A;

/**
 * Base binary name mapper implementation.
 */
public class BinaryBasicNameMapper implements BinaryNameMapper {
    /** Default use simple name flag setting. */
    public static final boolean DFLT_SIMPLE_NAME = false;

    /** */
    private boolean isSimpleName = DFLT_SIMPLE_NAME;

    /**
     * Default constructor.
     */
    public BinaryBasicNameMapper() {
        // No-op.
    }

    /**
     * @param isSimpleName Whether to use simple (no package) name of class or not.
     * <p>
     * Defaults to {@link #DFLT_SIMPLE_NAME}.
     */
    public BinaryBasicNameMapper(boolean isSimpleName) {
        this.isSimpleName = isSimpleName;
    }

    /**
     * Gets whether to use simple name of class or not.
     *
     * @return Whether to use simple name of class or not.
     */
    public boolean isSimpleName() {
        return isSimpleName;
    }

    /**
     * Sets whether to use simple name of class or not.
     *
     * @param isSimpleName Whether to use simple name of class or not.
     * @return {@code this} for chaining.
     */
    public BinaryBasicNameMapper setSimpleName(boolean isSimpleName) {
        this.isSimpleName = isSimpleName;

        return this;
    }

    /** {@inheritDoc} */
    @Override public String typeName(String clsName) {
        A.notNull(clsName, "clsName");

        return isSimpleName ? simpleName(clsName) : clsName;
    }

    /** {@inheritDoc} */
    @Override public String fieldName(String fieldName) {
        A.notNull(fieldName, "fieldName");

        return fieldName;
    }

    /**
     * @param clsName Class name.
     * @return Type name.
     */
    private static String simpleName(String clsName) {
        assert clsName != null;

        int idx = clsName.lastIndexOf('$');

        if (idx == clsName.length() - 1)
            // This is a regular (not inner) class name that ends with '$'. Common use case for Scala classes.
            idx = -1;
        else if (idx >= 0) {
            String typeName = clsName.substring(idx + 1);

            try {
                Integer.parseInt(typeName);

                // This is an anonymous class. Don't cut off enclosing class name for it.
                idx = -1;
            }
            catch (NumberFormatException ignore) {
                // This is a lambda class.
                if (clsName.indexOf("$$Lambda$") > 0)
                    idx = -1;
                else
                    return typeName;
            }
        }

        if (idx < 0)
            idx = clsName.lastIndexOf('+');   // .NET inner class.

        if (idx < 0)
            idx = clsName.lastIndexOf('.');

        return idx >= 0 ? clsName.substring(idx + 1) : clsName;
    }

    /** {@inheritDoc} */
    @Override public boolean equals(Object o) {
        if (this == o)
            return true;

        if (!(o instanceof BinaryBasicNameMapper))
            return false;

        BinaryBasicNameMapper mapper = (BinaryBasicNameMapper)o;

        if (isSimpleName != mapper.isSimpleName)
            return false;

        return true;
    }

    /** {@inheritDoc} */
    @Override public int hashCode() {
        return (isSimpleName ? 1 : 0);
    }

    /** {@inheritDoc} */
    @Override public String toString() {
        return "BinaryBaseNameMapper [isSimpleName=" + isSimpleName + ']';
    }
}
